### PR TITLE
Added solver part

### PR DESCRIPTION
**Pull request by bogdanprodanj**

### CODE REVIEW

File: relay/server/router.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ func (s *Server) buildRoutes() []route {
 func (s *Server) dAppRoutes() []route {
 	return []route{
 		{
-			name:        "SubmitUserOperation",
+			name:        "SubmitIntent",
```
Comment (i-Alex) [inline relay/server/router.go:68]: why was it renamed?
from the user perspective it is submit user operation

File: relay/server/solver_handler.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+		}
+		if err := conn.Reply(ctx, req.ID, response); err != nil {
+			logger.Error("error replying to client", "err", err, "reqID", req.ID, "caller", h.remoteAddress)
+		}
+	case methodSubscribe:
+		h.handleSubscribe(ctx, conn, req)
+	case methodUnsubscribe:
+		h.handleUnsubscribe(ctx, conn, req)
+	case methodSubmitSolverOperation:
+		h.handleSubmitIntentSolution(ctx, conn, req)
```
Comment (i-Alex) [inline relay/server/solver_handler.go:46]: here as well it should be "handleSubmitSolverOperation"

File: relay/server/solver.go [hunk 3]
```diff
@@ -19,6 +23,21 @@ var upgrader = websocket.Upgrader{
...
+	}
+
+	h := &wsConnHandler{
+		remoteAddress:       r.RemoteAddr,
+		intentService:       s.intentService,
+		subscriptionService: s.subscriptionService,
+	}
 
+	asyncHandler := jsonrpc2.AsyncHandler(h)
+	_ = jsonrpc2.NewConn(r.Context(), jsonrpc2_ws.NewObjectStream(connection), asyncHandler)
```
Comment (aanand1) [inline relay/server/solver.go:42]: Should pass some other context such as `context.Background` here. The reader's context `r.Context()` closes after this function causing the loop in `relay/server/solver_handler.go:handlerSubscriptionMessages` to terminate which prevents notifications from being sent.

File: relay/server/solver.go [hunk 4]
```diff
@@ -19,6 +23,21 @@ var upgrader = websocket.Upgrader{
 	},
 }
 
-func (s *Server) websocketSolver(w http.ResponseWriter, _ *http.Request) {
+func (s *Server) websocketSolver(w http.ResponseWriter, r *http.Request) {
+	connection, err := upgrader.Upgrade(w, r, nil)
```
Comment (aanand1) [inline relay/server/solver.go:27]: A close handler should be added for this connection which removes any associated subscription handlers. Without this, it tries (and fails) to send notifications to all previous connections which are now disconnected.

**Review by aanand1 — COMMENTED** [review]
I connected a `solver-1` to the `ws/solver` endpoint as defined in this PR, and and `solver-2` directly to the intent network. Upon sending an intent to the BDN, 

`solver-1` received - 
```
"{\"jsonrpc\":\"2.0\",\"method\":\"subscribe\",\"params\":{\"dappAddress\":\"0x54Dae1906e18e52410c26dAd4B4f14bace103E3E\",\"senderAddress\":\"0x54Dae1906e18e52410c26dAd4B4f14bace103E3E\",\"intentID\":\"837db67b-8e8b-4f95-8481-0e8c30e95ecc\",\"intent\":\"ZXlKamFHRnBia2xrSWpvaU1IZ3pPQ0lzSW5WelpYSlBjRWhoYzJnaU9pSXdlRGhoTVdKa1lXWm1Oalk0WVdJNU9HSmxNVEZtWkdaaFlXSTVaREkyT0RRd1pEZzBNMkpqTVROak16YzROamszTm1ReU5HUXlaVGMwT1RGaFpHVXhObVVpTENKMGJ5STZJakI0WkRjeVpEZ3lNV1JoT0RJNU5qUmpNRFUwTm1FMU5UQXhNelEzWVRNNU5UazRNRGhsTURjeVppSXNJbWRoY3lJNklqQjRaalF5TkRBaUxDSnRZWGhHWldWUVpYSkhZWE1pT2lJd2VEYzNNelU1TkRBd0lpd2laR1ZoWkd4cGJtVWlPaUl3ZURJNE1qWTFZallpTENKa1lYQndJam9pTUhnNU16QmhOR014TldVd05qZzFZalJtTm1JM09XTTJOamMzWXpCak56bGlPVE0yTkRCbE5XTTNJaXdpWTI5dWRISnZiQ0k2SWpCNE9ERmtNalV6TWpSaFlqWmhObUpqT1RJMFpqWXhaamhpTkRobFpqWTVPV1V4TkdabFkyUTVOQ0lzSW5aaGJIVmxJam9pTUhnd0lpd2laR0YwWVNJNklqQjRZekUwWXpreU1EUXdNREF3TURBd01EQXdNREF3TURBd01EQXdNREF3TURBd01EQXdNREF3TURBd01EQXdNREF3TURBd01EQXdNREF3TURBd01ERTVNV1EyTkdKaVlqQTROREUxTkRSak5ERTFNekprTkRRME5UUmtOR1l5WkRSa05HWTBNelJpTW1RME5qUTFORFUwTkRBd01EQXdNREF3TURBd01EQXdNREF3TURBd01EQXdNREF6T0RsbVpEazRNREF3TVRreFpEWTBZbUppTURnd01EQXdNREF3TlRBd01EQXdNV1V4T1ROaU1EWXlZalJsTnpsak9UbGpOelZrTkdKaVpUUmxNVFJqWmpJMk1HWTJNbVZpTkRGbU9XTXpZekZqTTJNMU5tUTNNbVpoTVRZeU1qRmpPVFUzTWpnNE1EbGxabU01T1RBek0yRTRNakV5TldKbU4yTmlZMlEwT0dJNU1EZzROVFl4Tm1VeU9HSXdNV0UxT1RZd00yTTRPREEwWVRFMVpUWmxNakZpTVdNd01EQXhNREF3TURBd01EQXdNREF5WldRMU56QXhNV1V3TURBd0lpd2labkp2YlNJNklqQjRZV016TURKaVkyRmpZakUzWlRVM09EazJZamRqWmpCaU4yVmpOakJqWTJFMU1HSmxaRFUzWVNKOQ==\",\"timestamp\":\"2024-09-09T10:19:33Z\"}}\n"
```

while `solver-2` received - 
```
"{\"method\":\"subscribe\",\"params\":{\"subscription\":\"0390d2bf-d205-4036-bf24-3c1f866df9f5\",\"result\":{\"dapp_address\":\"0x54Dae1906e18e52410c26dAd4B4f14bace103E3E\",\"sender_address\":\"0x54Dae1906e18e52410c26dAd4B4f14bace103E3E\",\"intent_id\":\"837db67b-8e8b-4f95-8481-0e8c30e95ecc\",\"intent\":\"eyJjaGFpbklkIjoiMHgzOCIsInVzZXJPcEhhc2giOiIweDhhMWJkYWZmNjY4YWI5OGJlMTFmZGZhYWI5ZDI2ODQwZDg0M2JjMTNjMzc4Njk3NmQyNGQyZTc0OTFhZGUxNmUiLCJ0byI6IjB4ZDcyZDgyMWRhODI5NjRjMDU0NmE1NTAxMzQ3YTM5NTk4MDhlMDcyZiIsImdhcyI6IjB4ZjQyNDAiLCJtYXhGZWVQZXJHYXMiOiIweDc3MzU5NDAwIiwiZGVhZGxpbmUiOiIweDI4MjY1YjYiLCJkYXBwIjoiMHg5MzBhNGMxNWUwNjg1YjRmNmI3OWM2Njc3YzBjNzliOTM2NDBlNWM3IiwiY29udHJvbCI6IjB4ODFkMjUzMjRhYjZhNmJjOTI0ZjYxZjhiNDhlZjY5OWUxNGZlY2Q5NCIsInZhbHVlIjoiMHgwIiwiZGF0YSI6IjB4YzE0YzkyMDQwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDE5MWQ2NGJiYjA4NDE1NDRjNDE1MzJkNDQ0NTRkNGYyZDRkNGY0MzRiMmQ0NjQ1NDU0NDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAzODlmZDk4MDAwMTkxZDY0YmJiMDgwMDAwMDAwNTAwMDAwMWUxOTNiMDYyYjRlNzljOTljNzVkNGJiZTRlMTRjZjI2MGY2MmViNDFmOWMzYzFjM2M1NmQ3MmZhMTYyMjFjOTU3Mjg4MDllZmM5OTAzM2E4MjEyNWJmN2NiY2Q0OGI5MDg4NTYxNmUyOGIwMWE1OTYwM2M4ODA0YTE1ZTZlMjFiMWMwMDAxMDAwMDAwMDAwMDAyZWQ1NzAxMWUwMDAwIiwiZnJvbSI6IjB4YWMzMDJiY2FjYjE3ZTU3ODk2YjdjZjBiN2VjNjBjY2E1MGJlZDU3YSJ9\",\"timestamp\":\"2024-09-09T10:19:33Z\"}},\"jsonrpc\":\"2.0\"}\n"
``` 

Why are the intent bytes in the two different? The `solver-2` is successfully able to parse the `intent` field into a `types.UserOperationPartialRaw` but `solver-1` isn't.

File: relay/server/solver_handler.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+	var p fastjson.Parser
+	v, err := p.ParseBytes(*req.Params)
+	if err != nil {
+		h.sendErrorMsg(ctx, jsonrpc2.CodeInvalidParams, fmt.Sprintf("failed to parse params: %v", err), conn, req.ID)
+		return
+	}
+
+	subscriptionType := v.GetStringBytes("subscription_type")
+	subscription, err := h.subscriptionService.Subscribe(h.remoteAddress, service.SubscriptionType(subscriptionType), conn)
```
Comment (aanand1) [inline relay/server/solver_handler.go:68]: Error handling similar to https://github.com/bloXroute-Labs/bdn-operations-relay/pull/7/files#r1750406337

File: relay/server/solver_handler.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+	var p fastjson.Parser
+	v, err := p.ParseBytes(*req.Params)
+	if err != nil {
+		h.sendErrorMsg(ctx, jsonrpc2.CodeInvalidParams, fmt.Sprintf("failed to parse params: %v", err), conn, req.ID)
+		return
+	}
+
+	subscriptionID := v.GetStringBytes("subscription_id")
+	err = h.subscriptionService.Unsubscribe(h.remoteAddress, string(subscriptionID))
```
Comment (aanand1) [inline relay/server/solver_handler.go:107]: Error handling similar to https://github.com/bloXroute-Labs/bdn-operations-relay/pull/7/files#r1750406337

File: relay/server/solver_handler.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+	v, err := p.ParseBytes(*req.Params)
+	if err != nil {
+		h.sendErrorMsg(ctx, jsonrpc2.CodeInvalidParams, fmt.Sprintf("failed to parse params: %v", err), conn, req.ID)
+		return
+	}
+
+	intentID := v.GetStringBytes("intent_id")
+	intentSolution := v.GetObject("intent_solution")
+
+	log.Debug("client submitted solver operation", "intent_id", string(intentID), "caller", h.remoteAddress)
```
[the remark this answers is not in the document]
Comment (bogdanprodanj) [reply relay/server/solver_handler.go:138]: The `*sdk.Client` takes care of these checks

[the remark this answers is not in the document]
Comment (aanand1) [reply relay/server/solver_handler.go:138]: if `intent_solution` field is missing or is not an object, `intentSolution.MarshalTo(nil)` this lines panics with a nil pointer

**Review by aanand1 — APPROVED** [review]

**Pull request merged**